File: packages/rust/slop-ai/src/transport/websocket.rs
//! WebSocket transport using tokio-tungstenite.
//!
//! ```no_run
//! use slop_ai::SlopServer;
//! use slop_ai::transport::websocket;
//!
//! #[tokio::main]
//! async fn main() {
//!     let slop = SlopServer::new("my-app", "My App");
//!     let handle = websocket::serve(&slop, "0.0.0.0:8765").await.unwrap();
//!     handle.await.unwrap();
//! }
//! ```

use std::collections::HashSet;
use std::net::{IpAddr, SocketAddr};
use std::sync::Arc;

use futures_util::{SinkExt, StreamExt};
use serde_json::Value;
use tokio::net::TcpListener;
use tokio::sync::mpsc;
use tokio::task::JoinHandle;
use tokio_tungstenite::tungstenite::handshake::server::{ErrorResponse, Request, Response};
use tokio_tungstenite::tungstenite::http::StatusCode;
use tokio_tungstenite::tungstenite::Message;

use crate::error::{Result, SlopError};
use crate::server::{Connection, SlopServer};

/// Decision returned by an [`Authenticator`].
///
/// Authentication happens during the HTTP upgrade, before any SLOP message
/// is delivered. Per spec/core/transport.md §Security, tokens MUST be
/// compared in constant time (see [`constant_time_eq`]) and MUST NOT be
/// logged.
pub type Authenticator =
    Arc<dyn Fn(&Request) -> std::result::Result<(), ErrorResponse> + Send + Sync>;

/// Options for [`serve_with_options`].
#[derive(Clone, Default)]
pub struct ServeOptions {
    /// Called for every upgrade request. Return `Ok(())` to accept, or an
    /// `ErrorResponse` (typically 401) to reject. If `None`, non-loopback
    /// upgrades are rejected by default.
    pub authenticate: Option<Authenticator>,
    /// Acceptable `Origin` values for browser clients. If empty, browser
    /// upgrades without a matching origin are rejected.
    pub allowed_origins: Vec<String>,
    /// Disable origin checking. Opt-in only, intended for local development.
    pub insecure_allow_all_origins: bool,
}

enum ConnMessage {
    Send(Value),
    Close,
}

struct ChannelConnection {
    tx: mpsc::UnboundedSender<ConnMessage>,
}

impl Connection for ChannelConnection {
    fn send(&self, message: &Value) -> Result<()> {
        self.tx
            .send(ConnMessage::Send(message.clone()))
            .map_err(|_| SlopError::Transport("connection closed".into()))
    }

    fn close(&self) -> Result<()> {
        let _ = self.tx.send(ConnMessage::Close);
        Ok(())
    }
}

fn error_response(status: StatusCode, body: &str) -> ErrorResponse {
    let mut resp = tokio_tungstenite::tungstenite::http::Response::new(Some(body.to_string()));
    *resp.status_mut() = status;
    resp
}

fn unauthorized() -> ErrorResponse {
    error_response(StatusCode::UNAUTHORIZED, "Unauthorized")
}

fn forbidden() -> ErrorResponse {
    error_response(StatusCode::FORBIDDEN, "Forbidden")
}

fn is_loopback(addr: &SocketAddr) -> bool {
    match addr.ip() {
        IpAddr::V4(ip) => ip.is_loopback(),
        IpAddr::V6(ip) => ip.is_loopback(),
    }
}

/// Start a SLOP WebSocket server with default secure settings.
///
/// Only loopback clients are accepted; use [`serve_with_options`] to
/// supply an authenticator for remote clients.
pub async fn serve(slop: &SlopServer, addr: &str) -> Result<JoinHandle<()>> {
    serve_with_options(slop, addr, ServeOptions::default()).await
}

/// Start a SLOP WebSocket server with the supplied authentication and
/// origin-check configuration.
pub async fn serve_with_options(
    slop: &SlopServer,
    addr: &str,
    opts: ServeOptions,
) -> Result<JoinHandle<()>> {
    let listener = TcpListener::bind(addr)
        .await
        .map_err(|e| SlopError::Transport(e.to_string()))?;

    let slop = slop.clone();
    let opts = Arc::new(opts);

    let handle = tokio::spawn(async move {
        while let Ok((stream, peer)) = listener.accept().await {
            let slop = slop.clone();
            let opts = opts.clone();
            tokio::spawn(async move {
                let allowed: HashSet<String> = opts.allowed_origins.iter().cloned().collect();
                let insecure = opts.insecure_allow_all_origins;
                let authenticate = opts.authenticate.clone();
                let peer_loopback = is_loopback(&peer);

                let callback = |req: &Request, response: Response| -> std::result::Result<Response, ErrorResponse> {
                    // Origin allowlist (only applies when client sent Origin).
                    if !insecure {
                        if let Some(origin) = req.headers().get("origin") {
                            let ok = origin
                                .to_str()
                                .ok()
                                .map(|s| allowed.contains(s))
                                .unwrap_or(false);
                            if !ok {
                                return Err(forbidden());
                            }
                        }
                    }

                    if let Some(ref auth) = authenticate {
                        auth(req)?;
                    } else if !peer_loopback {
                        eprintln!(
                            "[slop] refusing non-loopback WebSocket upgrade: no authenticate hook configured. \
                             See spec/core/transport.md §Security considerations."
                        );
                        return Err(unauthorized());
                    }

                    Ok(response)
                };

                let ws_stream = match tokio_tungstenite::accept_hdr_async(stream, callback).await {
                    Ok(ws) => ws,
                    Err(_) => return,
                };

                let (mut sender, mut receiver) = ws_stream.split();
                let (tx, mut rx) = mpsc::unbounded_channel::<ConnMessage>();
                let conn: Arc<dyn Connection> = Arc::new(ChannelConnection { tx });

                // Spawn a writer task that drains the channel into the WS sink
                tokio::spawn(async move {
                    while let Some(msg) = rx.recv().await {
                        match msg {
                            ConnMessage::Send(val) => {
                                let json = serde_json::to_string(&val).unwrap_or_default();
                                if sender.send(Message::Text(json.into())).await.is_err() {
                                    break;
                                }
                            }
                            ConnMessage::Close => {
                                let _ = sender.send(Message::Close(None)).await;
                                break;
                            }
                        }
                    }
                });

                slop.handle_connection(conn.clone());

                while let Some(Ok(msg)) = receiver.next().await {
                    if let Message::Text(text) = msg {
                        if let Ok(parsed) = serde_json::from_str::<Value>(&text) {
                            slop.handle_message(&conn, &parsed);
                        }
                    }
                }

                slop.handle_disconnect(&conn);
            });
        }
    });

    Ok(handle)
}

/// Constant-time comparison helper for bearer-token equality checks.
pub fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff: u8 = 0;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}

File: packages/python/slop-ai/src/slop_ai/transports/websocket.py
"""WebSocket transport for SLOP using the ``websockets`` library.

Usage::

    from slop_ai import SlopServer
    from slop_ai.transports.websocket import serve

    slop = SlopServer("my-app", "My App")

    async def main():
        server = await serve(slop, host="0.0.0.0", port=8765)
        await server.wait_closed()
"""

from __future__ import annotations

import asyncio
import inspect
import ipaddress
import json
import logging
from typing import Any, Awaitable, Callable, Iterable

try:
    import websockets
    from websockets.asyncio.server import Server, ServerConnection, serve as ws_serve
except ImportError as e:
    raise ImportError(
        "websockets is required for the WebSocket transport. "
        "Install it with: pip install slop-ai[websocket]"
    ) from e

from slop_ai.server import SlopServer

_log = logging.getLogger(__name__)

# Authenticate hook: receives the incoming Request and returns True to accept.
# May be sync or async; raising is treated the same as returning False.
Authenticator = Callable[[Any], "bool | Awaitable[bool]"]


def _is_loopback_remote(ws: ServerConnection) -> bool:
    peer = getattr(ws, "remote_address", None)
    if not peer:
        return False
    host = peer[0] if isinstance(peer, tuple) else str(peer)
    try:
        return ipaddress.ip_address(host).is_loopback
    except ValueError:
        return host in ("localhost", "::1")


async def serve(
    slop: SlopServer,
    host: str = "localhost",
    port: int = 8765,
    path: str = "/slop",
    *,
    authenticate: Authenticator | None = None,
    allowed_origins: Iterable[str] | None = None,
) -> Server:
    """Start a standalone SLOP WebSocket server.

    Per spec/core/transport.md §Security considerations, non-loopback upgrades
    without an ``authenticate`` hook are rejected with 401. Set
    ``authenticate=lambda req: True`` to opt out explicitly (not recommended).

    Returns a ``websockets.Server`` — call ``await server.wait_closed()``
    to block until shutdown.
    """

    origins_set = set(allowed_origins) if allowed_origins is not None else None

    async def handler(ws: ServerConnection) -> None:
        if ws.request is None:
            await ws.close(4000, "Bad Request")
            return

        if ws.request.path != path:
            await ws.close(4004, f"Not found: {ws.request.path}")
            return

        # Origin allowlist (only applies for browser clients that send Origin).
        origin = ws.request.headers.get("Origin")
        if origin is not None and origins_set is not None and origin not in origins_set:
            await ws.close(4003, "Forbidden")
            return

        # Authentication. Default-deny non-loopback when no hook is configured.
        if authenticate is not None:
            try:
                result = authenticate(ws.request)
                if inspect.isawaitable(result):
                    result = await result
                if not result:
                    await ws.close(4401, "Unauthorized")
                    return
            except Exception:
                await ws.close(4401, "Unauthorized")
                return
        elif not _is_loopback_remote(ws):
            _log.warning(
                "[slop] refusing non-loopback WebSocket upgrade: no authenticate hook configured. "
                "See spec/core/transport.md §Security considerations."
            )
            await ws.close(4401, "Unauthorized")
            return

        conn = _WsConnection(ws)
        slop.handle_connection(conn)

        try:
            async for raw in ws:
                try:
                    msg = json.loads(raw)
                    await slop.handle_message(conn, msg)
                except json.JSONDecodeError:
                    pass
        finally:
            slop.handle_disconnect(conn)

    server = await ws_serve(handler, host, port)
    return server


class _WsConnection:
    """Wraps a websockets connection as a SLOP Connection."""

    __slots__ = ("_ws",)

    def __init__(self, ws: ServerConnection) -> None:
        self._ws = ws

    def send(self, message: dict[str, Any]) -> None:
        asyncio.ensure_future(self._ws.send(json.dumps(message)))

    def close(self) -> None:
        asyncio.ensure_future(self._ws.close())
